### PR TITLE
[WEJBHTTP-140] Second round of refactorings in TXN part - focus on client side of the protocol

### PR DESCRIPTION
**Pull request by ropalka**
https://issues.redhat.com/browse/WEJBHTTP-140

### CODE REVIEW

File: transaction/src/main/java/org/wildfly/httpclient/transaction/ClientHandlers.java [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+import org.wildfly.httpclient.common.NoFlushByteOutput;
+import org.xnio.IoUtils;
+
+import javax.transaction.xa.Xid;
+import java.io.Closeable;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+
```
Comment (rachmatowicz) [inline transaction/src/main/java/org/wildfly/httpclient/transaction/ClientHandlers.java:41]: Comments would help here to describe the overall purpose of this class in the request/response interaction, something like:
"This class defines helper methods for marshaling common transaction invocation request parameter types, as well as handling the common transaction invocation response result types"

File: transaction/src/main/java/org/wildfly/httpclient/transaction/HttpRemoteTransactionHandle.java [hunk 2]
```diff
@@ -143,26 +134,18 @@ public void rollback() throws SecurityException, SystemException {
...
-                    marshaller.finish();
-                }
-            }, (input, response, closeable) -> {
-                try {
-                    result.complete(null);
-                } finally {
-                    IoUtils.safeClose(closeable);
-                }
-            }, result::completeExceptionally, null, null);
+            final CompletableFuture<Void> result = new CompletableFuture<>();
```
Comment (rachmatowicz) [inline transaction/src/main/java/org/wildfly/httpclient/transaction/HttpRemoteTransactionHandle.java:142]: This looks much cleaner!

**Review by rachmatowicz — APPROVED** [review]
Had a suggestion for using a switch statement and some additional comments in places, but looks good.

File: transaction/src/main/java/org/wildfly/httpclient/transaction/RequestType.java [hunk 3]
```diff
@@ -20,15 +20,6 @@
 
 import static io.undertow.util.Methods.GET;
 import static io.undertow.util.Methods.POST;
-import static org.wildfly.httpclient.transaction.TransactionConstants.UT_BEGIN_PATH;
```
[the remark this answers is not in the document]
Comment (ropalka) [reply transaction/src/main/java/org/wildfly/httpclient/transaction/RequestType.java:23]: Correct.

File: transaction/src/main/java/org/wildfly/httpclient/transaction/HttpRemoteTransactionHandle.java [hunk 4]
```diff
@@ -84,16 +88,11 @@ public void commit() throws RollbackException, HeuristicMixedException, Heuristi
 
             targetContext.sendRequest(request, sslContext, authenticationConfiguration, output -> {
                 Marshaller marshaller = targetContext.getHttpMarshallerFactory(request).createMarshaller();
-                marshaller.start(Marshalling.createByteOutput(output));
```
[the remark this answers is not in the document]
Comment (ropalka) [reply transaction/src/main/java/org/wildfly/httpclient/transaction/HttpRemoteTransactionHandle.java:87]: Thanks.

File: transaction/src/main/java/org/wildfly/httpclient/transaction/HttpRemoteTransactionService.java [hunk 5]
```diff
@@ -85,18 +84,35 @@ public HttpRemoteTransactionService(LocalTransactionContext transactionContext,
...
+        registerHandler(routingHandler, XA_RECOVER);
+        registerHandler(routingHandler, XA_ROLLBACK);
         return httpServiceConfig.wrap(new BlockingHandler(new ElytronIdentityHandler(routingHandler)));
     }
 
+    private void registerHandler(final RoutingHandler routingHandler, final RequestType requestType) {
+        routingHandler.add(requestType.getMethod(), requestType.getPath(), newInvocationHandler(requestType));
+    }
+
+    private HttpHandler newInvocationHandler(final RequestType requestType) {
```
[the remark this answers is not in the document]
Comment (ropalka) [reply transaction/src/main/java/org/wildfly/httpclient/transaction/HttpRemoteTransactionService.java:93]: Yes, it is used. It was introduced in next commits.

File: transaction/src/main/java/org/wildfly/httpclient/transaction/HttpRemoteTransactionHandle.java (hunk 2 above)
Comment (ropalka) [reply transaction/src/main/java/org/wildfly/httpclient/transaction/HttpRemoteTransactionHandle.java:142]: Thanks.

File: transaction/src/main/java/org/wildfly/httpclient/transaction/ClientHandlers.java (hunk 1 above)
Comment (ropalka) [reply transaction/src/main/java/org/wildfly/httpclient/transaction/ClientHandlers.java:41]: How about: "Provides factory methods for client-side handlers of the TXN over HTTP protocol." ?

**Review by fl4via — APPROVED** [review]

**Pull request merged**